### PR TITLE
Move SolrCloud to beanstalk

### LINKED ISSUE #126: Move SolrCloud deployment from ECS to Beanstalk
Issue #125 moves Zookeeper to Beanstalk, this work moves SolrCloud and moves the last component in the Hyku stack out of ECS.

Need to ensure that each of the Solr nodes is able communicate with each of the Zookeeper nodes directly (not via ELB.)

Alter the Docker image we deploy to query the exhibitor API for the zk nodes on service startup

### PR DESCRIPTION
**Pull request by eefahy**
One existing issue is our inability to tear down the stack cleanly due to a bug in cloudformation. As a result cloudformation is not able to remove a network interface it builds for the zookeeper-dns lambda script and therefore fails to delete the securitygroups stack. The fix is to manually remove the network interface when the securitygroups stack is deleting. :(

Closes #126 #125 #105

### CODE REVIEW

**Review by dbernstein — COMMENTED** [review]
This thing is so intense.  A couple of questions:  
1. why is assets/solr/solr-v1.zip stored in github? Shouldn't that live in a public s3 bucket somewhere?
2. I'm assuming I need to merge the changes in params/defaults.json into my params/private.json file if I can have any hope of this working?

**Comment by dbernstein** [comment]
@eefahy :  can you tell me what the criteria would be for a successful test beyond firing up the stack and deleting it?

File: params/defaults.json [hunk 1]
```diff
@@ -109,15 +109,15 @@
...
   {
-    "ParameterKey": "SolrCloudDockerImage",
-    "ParameterValue": "hybox/solr-ecs:6"
+    "ParameterKey": "SolrHealthReportingSystemType",
+    "ParameterValue": "enhanced"
   },
   {
-    "ParameterKey": "SolrCloudDataVolumePath",
-    "ParameterValue": "/var/lib/docker/vfs/dir/solr-data"
+    "ParameterKey": "SolrS3Key",
```
Comment (bbranan) [inline params/defaults.json]: What is actually stored in the Solr and Zookeeper S3 buckets? If these need to be created prior to building the stack, we'll need to add that to the readme. Also, can these two be the same bucket, or do they need to be distinct?

**Comment by bbranan** [comment]
Changes needed in the readme (based on the updates in this PR):
- Any file under /params/private/ is now ignored, not just private.json
- Description of how to create the Solr deploy bundle
- Description of how to create the Zookeeper deploy bundle
- Description of how to create the Fcrepo deploy bundle, with a note about how to change the Fcrepo home path if that is desired

File: templates/solr.yaml [hunk 2]
```diff
@@ -1,12 +1,33 @@
 AWSTemplateFormatVersion: '2010-09-09'
-Description: Hyku SolrCloud stack using ECS
+Description: Hyku solrcloud stack
```
Comment (bbranan) [inline templates/solr.yaml]: Perhaps "Hyku Solr instance", given that there is now a distinct SolrCloud template?

File: templates/solr.yaml [hunk 3]
```diff
@@ -1,12 +1,33 @@
...
+    Description: Maximum number of instances
+  InstanceType:
+    Type: String
+    Description: The EC2 instance type
+  HealthReportingSystemType:
+    Type: String
+    Description: Health reporting system
+  S3Bucket:
+    Type: String
+    Description: Bucket with ZK source bundle
```
Comment (bbranan) [inline templates/solr.yaml]: Bucket with Solr source bundle

File: templates/solr.yaml [hunk 4]
```diff
@@ -1,12 +1,33 @@
...
+  S3Key:
+    Type: String
+    Description: Key for ZK source bundle
```
Comment (bbranan) [inline templates/solr.yaml]: Key for Solr source bundle

File: templates/zookeeper-dns.yaml [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    Type: List<AWS::EC2::SecurityGroup::Id>
+    Description: Private subnets for zk stack
+Resources:
+  BeanstalkStack:
+    Type: Custom::BeanstalkStack
+    Properties:
+      ServiceToken: !GetAtt [BeanstalkStackOutputs, Arn]
+      EnvironmentName: !Ref ZookeeperEnvironmentName
+  BeanstalkStackOutputs:
+    Type: AWS::Lambda::Function
```
Comment (bbranan) [inline templates/zookeeper-dns.yaml]: Would be helpful to have a comment here describing the purpose of this Lambda function

File: templates/zookeeper-dns.yaml [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+      FunctionName: !GetAtt [UpsertRoute53Record, Arn]
+      Action: lambda:InvokeFunction
+      Principal: events.amazonaws.com
+  InitializeUpsertLambda:
+    Type: Custom::InitializeUpsertLambda
+    Properties:
+      ServiceToken: !GetAtt [UpsertRoute53Record, Arn]
+      ZookeeperASGName: !GetAtt [BeanstalkStack, AutoScalingGroupName]
+  UpsertRoute53Record:
+    Type: AWS::Lambda::Function
```
Comment (bbranan) [inline templates/zookeeper-dns.yaml]: Would be helpful to have a comment here describing the purpose of this Lambda function

**Review by bbranan — COMMENTED** [review]
Changes look good to me, I just noted a few places where descriptions or comments could be updated. Removing the solr-v1.zip file from the commit is the only real blocker for merging.

**Comment by eefahy** [comment]
@bbranan I've added commits per your review

**Review by dbernstein — APPROVED** [review]
These look good to me.

**Pull request merged**